### ISSUE TITLE
HD Mod supported features

### ISSUE THREAD

**Issue by krs0** [vcmi/vcmi#1613]
This is a compilation of all stuff HD mod offers. A comprehensive list, that can shows what VCMI already covers and what still needs to be developed.

This list was compiled from HD mod's list of features available  [here](https://sites.google.com/site/heroes3hd/eng)

**Graphics**
- [ ] Allow resizing of game window via mouse dragging
- [ ] Scaling/stretching to the desired resolution
- [ ] Support different resolutions (Up to 4000x4000)
- [ ] Improved cursors (System Cursors for Windows Vista or higher with scaling)

**Sounds**
- [ ] Do not force player to wait for sounds to finish (EG: do not wait for start combat music to finish)
- [x] Game music in 44 kHz stereo (VCMI already plays music with resampling to 44 kHz)

**Adventure Screen**
- [ ] Extra adventure screen buttons (EG: for adventure options)
- [ ] Extended hero/towns lists, with scroll buttons, mouse scroll support, sorting
- [ ] Status Window does not cycle through Hero Info/Day/Allies Info on left click, but shows only selected hero/town.
- [ ] Status Window is fully functional (Quick Army management, etc)
<img src="https://user-images.githubusercontent.com/15194321/221404693-9f225e5e-1702-43ca-a571-342f8cfc445a.png" width=25% height=25%>

- [ ] Global hotkeys for Market, Thieves guild, hero costumes - #1382
- [ ] Quick menu (show map grid, market, thieves' guild, quick combat, DD area, replay my turn)
<img src="https://user-images.githubusercontent.com/15194321/221404173-0bc7a1d6-6b8e-4e19-8370-a026c42ecc1d.png" width=10% height=10%>

- [x] Hero movement points information (Toggle option in settings: Always show movement cost)
<img src="https://user-images.githubusercontent.com/15194321/221404565-cc5bfe24-66f7-44c9-931c-4f51ac19baeb.png" width=50% height=50%>

**Other Quality of life improvements**
- [x] Remove unnecessary confirmations for all original... confirmation windows popups.
<img src="https://user-images.githubusercontent.com/15194321/221404794-8539e5bd-b4da-4160-942f-2f114d805c3b.png" width=25% height=25%>

- [ ] Turn timer should be always visible

**Hero Window**
<img src="https://user-images.githubusercontent.com/15194321/221404915-a4c15893-8a4e-4e1b-90f3-f1abb620a62e.png" width=50% height=50%>
- [ ] Quick Army Management
- [ ] Quick Artifact Management
- [ ] Artifact Backpack (Add a separate window for hero's artifact backpack)
- [ ] Artifact sets (costumes) (Allow players to save and switch artifact configurations)

**Heroes Meeting Screen**
<img src="https://user-images.githubusercontent.com/15194321/221405022-0725391e-9b6f-408f-8a88-874aa79a4e65.png" width=60% height=60%>
- [ ] Quick Army management
- [ ] Quick Artifact management
- [ ] Extra button for closing window and focusing on chosen hero
- [ ] Artifact backpack management
- [ ] Artifact sets (costumes) - #999

**Combat Screen**
<img src="https://user-images.githubusercontent.com/15194321/221405115-7b1dd5e6-4336-46f8-9de8-8f3639aa159d.png" width=70% height=70%>
- [ ] Turn order bar (Add a bar that displays creature turn order, like the initiative bar in H5)
- [ ] Separate Log and Status Bars - #925 
- [x] Improve creature selection in tactics mode (EG: Select creatures with mouse) 
- [ ] Add more shortcuts (Load, Finish with autocombat, etc)
- [ ] Extra buttons for creature actions (harpy hags, magogs, etc)
- [ ]  https://github.com/vcmi/vcmi/issues/1666
<img src="https://user-images.githubusercontent.com/15194321/224374208-fc501875-4942-4507-844a-d5130b9a17e5.png" width=30% height=30%>  

**Combat Options dialog**
- [ ] Extra options in "Combat Options dialog", (game save, load, exit)

**Spell Book**
<img src="https://user-images.githubusercontent.com/15194321/221405242-bdd70feb-a0d6-443f-af88-cf1087240be7.png" width=70% height=70%>
- [ ] Bigger Spell book (Fit more spells on a page)
- [ ] Better Spell information (Add Spell damage calculations)

**Town Portal Dialog**
- [ ] Show Town icons
- [ ] Selected town should also be selected on minimap
- [ ] Inform player if hero occupies town
- [ ] List towns in same order as on Adventure Screen

**Town Screen**
- [ ] Quick Army management available
- [ ] - [SPACE] hotkey not only swap heroes but move hero to garrison and back
- [ ] Quick recruit (Add a central place for recruiting creatures, recruit all at once.. .etc)
- [ ] Chat in town screen
- [ ] Show creature growth / Available creatures to recruit
- [x] Extended creature growth information (show what bonuses come from buildings / artifacts...)

**Scenario Manager**
- [ ] Allow subfolders of Saves folder
- [ ] Games are saved in a separate folder with the scenario name if a tournament option ‘save all each day’ is on.
- [ ] Automatic saves before each battle
- [ ] Full information on starting hero
- [ ] Easier town selection (window with icons of all towns)
- [ ] Easier hero selection (window with icons of all heroes)
- [ ] Right click on map name to show related file name (added to list by Dydzio 11.03.2023)

**Random Map Generator (RMG)**
- [x] Choose template option (allow switching template from in-game)
- [x] Choose road types (User can select what road types should be used for RMG)
- [ ] Persistent map settings between game sessions (selected template, selected map, difficulty)
- [x] New popup window for easy team selection

**Miscellaneous**
- [ ] Invite heroes in taverns (In Single player add an option to allow recruiting of desired hero.)
- [ ] Easy choosing for a secondary skill in Level Up Dialog by double clicking it.
- [ ] Quality of life global shortcuts

**Multiplayer**
- [ ] Simultaneous Turns
- [ ] Online Lobby
- [ ] Protect multiplayer game and a RMG map with a password.
- [ ] Chess Timmer
- [ ] Replay combat option for Single Player

**Comment by rilian-la-te** [comment]
@krs0 movement points is done in develop AFAIK.

**Comment by Nordsoft91** [comment]
RMG features implemented - activate "extended RMG" submod from vcmi extras

**Comment by Nordsoft91** [comment]
"Extra buttons for creature actions (harpy hags, magogs, etc)" - done with "battlefield actions" submod from vcmi extras

**Comment by rilian-la-te** [comment]
@Nordsoft91 please, submit this mod to vcmi-1.2 branch)

**Comment by kambala-decapitator** [comment]
>  Improve creature selection in tactics mode (EG: Select creatures with mouse)

we already have this

> Inform player if hero occupies town

currently such towns are simply not listed if I'm not mistaken

> List towns in same order as on Adventure Screen

isn't it already the case?

**Comment by krs0** [comment]
Maybe there is a misconception on what this epic's purpose is....
I will clarify in description.
This is a compilation of all stuff HD mod offers. A comprehensive list, that can shows what VCMI already covers and what still needs to be developed.
We should test and check the boxes one by one if they are available, list deviations, provide info when they make no sesnse in VCMI context.

**Comment by psyhlo** [comment]
HD Mod offers scaling/stretching to the desired resolution

**Comment by rilian-la-te** [comment]
@krs0 you should check something than implemented, I think)

**Comment by krs0** [comment]
> @krs0 you should check something than implemented, I think)

? Can you rephrase please. I did not catch that :)

**Comment by IvanSavenko** [comment]
List of HD features (using issue list) that in my opinion should not be implemented at all or should be only implemented after careful considerations & potential redesign:

- Quick menu (show map grid, market, thieves' guild, quick combat, DD area, replay my turn)
-> should be done either via hotkeys or with menu accessible with left click, e.g. new button in UI.
- Artifact sets (costumes) 
-> needs proper UI design
- Artifact Backpack 
-> needs proper UI design
- Extra button for closing hero exchange window and focusing on chosen hero
-> very situational and increases cognitive load on already extremely heavy UI window
- Allow subfolders of Saves folder
-> very limited use-case
- tournament option ‘save all each day’ 
-> should be done differently
- Invite heroes in taverns 
-> against it, balance changing feature and not in a good way
- Protect multiplayer game and a RMG map with a password.
-> wrong & insecure approach
- Replay combat option for Single Player
-> should be limited to replaying AI quick combat results

**Comment by krs0** [comment]
Extra button for closing hero exchange window and focusing on chosen hero
-> very situational and increases cognitive load on already extremely heavy UI window

When you do hero chains this saves time. You simply exit the menu with selecting the right hero. Personally got so used to it I use it 100% even when no chains involved.
It is 2 buttons in the end.

................

Allow subfolders of Saves folder
-> very limited use-case

100s of saves need to be organized somehow. EG using 1 folder for each map.

Invite heroes in taverns
-> against it, balance changing feature and not in a good way

This is a single player only option. Quite usefull in some cases... Came as a feature because lots of players just restarted the scenario multiple times until satisfied with results. So if restarts produce same result I see no problem in having it as an option. low prio thouh.

...................

Replay combat option for Single Player
-> should be limited to replaying AI quick combat results

How do players do it now? They save the game before battle and restart if they want to replay battle (with rearranged army if needed)
Different people enjoy the game differently, I see this as a must. I played special maps that force you to do a battle at the very limit, could be done only via save/load multiple times

**Comment by IvanSavenko** [comment]
> Came as a feature because lots of players just restarted the scenario multiple times until satisfied with results

In that case should we also
- allow players to select starting artifact? After all, you can restart as long as you wish till you get one you want
- allow players to select starting armies?
- allow players to select random artifacts/resources in starting zone?

There is nothing good in allowing players to bypass randomization.

> How do players do it now? They save the game before battle and restart if they want to replay battle (with rearranged army if needed)

So what? This is not behavior that should be encouraged/prioritized. Using same logic - perhaps we should allow to rollback combat to any suitable position?
Enemy creature got lucky strike? Rollback combat to before its turn and cast blind/slow
Your creature failed to deal enough damage to kill last dragon and got killed? Reroll to just before this point.

Heck, I'd rather support implementation of single-save/no reload/Ironman option for hardcore players rather than implement savescumming features

**Comment by SoundSSGood** [comment]
> * Invite heroes in taverns

Please don't implement this. Very bad decision in WOG

**Comment by krs0** [comment]
HD stuff should be prioritized a little.
Invite heroes is a very very low prio, if at all...

@SoundSSGood can you explain how is this feature impacting WoG?

@IvanSavenko saving before battles was always a thing in H3. It is only for single player, many do it, I would not force players to play in a certain way, on the contrary I would like support for that like I have in HD mod. I saw no complaints about this HD mod feature.

**Comment by IvanSavenko** [comment]
> HD stuff should be prioritized a little. Invite heroes is a very very low prio, if at all...

Invite hero should not be in. Maybe if at some we implement scripting. But definitely not something that should be in vcmi codebase.

> @IvanSavenko saving before battles was always a thing in H3.

Yes. Clicking on save button before potentially dangerous battle, selecting save, then going to main menu and loading game. Not 1-button restart on battle result screen.
Save-scumming does *not* improves the game.

**Comment by SoundSSGood** [comment]
> can you explain how is this feature impacting WoG?

WOG implemented a lot of features without caring about balance. Invite heroes is one of them.
Perhaps for compatibility with WOG, it should be implemented in the future. But as @IvanSavenko said, not in the VCMI codebase.

**Comment by rilian-la-te** [comment]
@SoundSSGood when LUA gets its golden hour, then we can add support of arbitrary city scripts. And somebody can implement invite heroes mod. It should not be core feature.

**Comment by dydzio0614** [comment]
> Heck, I'd rather support implementation of single-save/no reload/Ironman option for hardcore players rather than implement savescumming features

this sounds fairly cool, assuming it gets connected with decent high scores menu and score calculation algorithm

though this is far from wanted stuff for now, for either players or modders

**Comment by dydzio0614** [comment]
So to more or less sum up player & discord & slack discussions/feedback:
there tend to be varying opinions for last 5 items on ivan's list (except password protection, assuming we manage to implement solution that is as good):

- saving everyday and save folders are somewhat related when used in multiplayer environment, player want wasy saves grouping and potentially easily doing that also in file explorer to send saves from arbitrary games without much effort etc. There are various reasons for saving everyday - studying matches, testing another approaches to past turns after match finishes, checking for cheating (this one should be not needed in VCMI)
- hero invite has use in singleplayer tournaments and testing how lucky hero rolls work in practice in simulated competitive early game environment via singleplayer, etc. My previous comment gives additional points. IMO this thing should be made possible in VCMI, preferably via script.
- infinite replay has various small use cases and I definitely cannot list all of them - among them there are features such as hardcore maps, singleplayer tournaments, practicing AI cheesing and finding new competitive tactics where player is overwhelmed in particular fights (for example creature bank). Cheesing also got its audience, same as there are people who play with cheats, or those who solo play hotseat without AI. Same as with hero invite - there are some people who treat this as basic feature as well because they only played with HD mod. It's also not like we can't make this thing existing but disabled by default. I wouldn't want to see some people being disappointed in VCMI because we decline to implement something, especially while it is about things that are nowhere near our current priority list anyway.

**Comment by rilian-la-te** [comment]
@krs0 infobar is merged, you can check this.

**Comment by IvanSavenko** [comment]
Probably closing - moved to Github Project
https://github.com/orgs/vcmi/projects/2

**Comment by Shatur** [comment]
It says that `Scaling/stretching to the desired resolution` is supported partially now. Is scaling works now?
